### ISSUE TITLE
"Open in App" button functionality is blocked by browser

### ISSUE THREAD

**Issue by memo33** [memo33/sc4pac-gui#15]
The ["Open in App"](https://github.com/memo33/sc4pac-tools/pull/21) button on the [website](https://memo33.github.io/sc4pac/channel/?pkg=memo:submenus-dll) does not work in some web browsers, as the POST request to `localhost:51515` may be blocked by an ad blocker.

Workaround: Users would need to disable the ad blocker for the domain `memo33.github.io`. For example, by pressing the Shields icon in Brave's address bar.

This is quite unfortunate.

A proper solution would involve registering a new protocol handler for a custom URI scheme. This is heavily platform-dependent, so it seems like a lot of effort to get this to work.
* Windows: requires editing the registry
* macOS: requires creating an app bundle for the application ([example](https://relentlesscoding.com/posts/create-macos-app-bundle-from-script/)) (not sure if practical without apple developer license)
* Linux: not a concern for now

**Comment by memo33** [comment]
I've done a bit more research. It looks like it's not necessarily ad-blockers that block localhost access, but web browsers themselves:
- Safari (and WebKit browsers)
- [Brave](https://brave.com/privacy-updates/27-localhost-permission/) browser

I've tried to disable this in Brave, but wasn't quite successful.
- First, enable an experimental setting under `brave://flags/#brave-localhost-access-permission` for localhost access
- Then, when you visit a website that accesses localhost, a popup should be displayed asking for permission. (This seems to be broken – I've never managed to get the popup to display. Maybe websites need to request a specific permission for this somehow?)
- Also, under `brave://settings/content` > `Permissions` > `Additional permissions` > `Localhost access`, it's possible to whitelist the site `memo33.github.io` directly. (This seems to work, but not in Private mode – Why…!?)

The other workaround is:
- Disable the experimental setting under `brave://flags/#brave-localhost-access-permission` for localhost access. (This is the default.)
- Disable "Shields" for the site. (This has the side effect of allowing all other trackers as well).

**Comment by memo33** [comment]
For Brave, the main background info I found are [a blogpost](https://brave.com/privacy-updates/27-localhost-permission/) and the PR that implemented the localhost access: `https://github.com/brave/brave-browser/issues/27346`. Official developer documentation seems to be scarce for Brave in general. @sebamarynissen

**Comment by sebamarynissen** [comment]
Alright, thanks for that information. I kind of get the point of why Brave blocks calls to localhost. It makes sense in their philosophy of limiting tracking as much as possible.

The ideal solution seems to be to register a custom `sc4pac://` protocol, and I've managed to get a PoC working with Electron. Electron [provides an easy way to register an application as a protocol handler](https://www.electronjs.org/docs/latest/tutorial/launch-app-from-url-in-another-app), without having to edit the registry manually and take care of cross-platform differences. It should work on Mac too (though not tested) if you have packaged the app, no Apple developer license needed.

I've managed to run the webapp inside an Electron application (see code below). This allows me to include a link to `sc4pac://sc4pac.sebamarynissen.dev/?pkg=jasoncw:xenon-center` and then have my browser ask me to open the Electron application. Then I parse the url passed in, and call the server on `/packages.open`. It works beautifully, because it even opens the application when it's not running yet. 

There's still a lot of polishing needed of course - for example the case of opening multiple packages from multiple channels - but for me it shows that the `sc4pac://` protocol is the way to go. I'm not sure how familiar you are with Electron, so ideally the protocol handler can be registered in Flutter as well, but if not feasible, then I think I would be able to setup an action that will package the Flutter webapp in an Electron app. Let me know what your thoughts are.

```js
import { app, BrowserWindow, dialog } from 'electron/main';
import cp from 'node:child_process';
import path from 'node:path';

if (process.defaultApp) {
  if (process.argv.length >= 2) {
    app.setAsDefaultProtocolClient('sc4pac', process.execPath, [path.resolve(process.argv[1])]);
  } else {
    app.setAsDefaultProtocolClient('sc4pac');
  }
}

// Ensure only a single app instance can be running.
const gotTheLock = app.requestSingleInstanceLock();
if (!gotTheLock) {
  app.quit();
} else {
  app.on('second-instance', (event, commandLine, cwd) => {
    let url = commandLine.pop();
    openUrl(url);
  });
}

// Launches the sc4pac server.
async function sc4pac() {
  let child = cp.spawn('cli\\sc4pac.bat', [
      'server',
      '--port 51515',
      '--web-app-dir webapp',
    ], {
    shell: true,
  });
}

Promise.all([
  app.whenReady(),
  sc4pac(),
]).then(() => {
  createWindow();
});

let mainWindow;
function createWindow() {
  mainWindow = new BrowserWindow({
    autoHideMenuBar: true,
    width: 800,
    height: 600,
  });
  mainWindow.loadURL('http://localhost:51515/webapp');
}

async function openUrl(raw) {
  if (!mainWindow) createWindow();
  if (mainWindow.isMinimized()) mainWindow.restore();
  mainWindow.focus();
  let url = new URL(raw);
  console.log(url);
  let channelUrl = `https://${url.host}${url.pathname}`.replace(/\/$/, '')+'/';
  let payload = [
    {
      package: url.searchParams.get('pkg'),
      channelUrl,
    },
  ];
  let body = JSON.stringify(payload);
  await fetch('http://localhost:51515/packages.open', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': body.length,
    },
    body,
  });
}

```

**Comment by sebamarynissen** [comment]
Managed to package something for Windows and Linux (available here: https://github.com/sebamarynissen/sc4pac-protocol/actions/runs/12611055744). I just realized though that for Linux it won't work because I'm launching the sc4pac server with `sc4pac.bat`, which isn't going to work on Linux. I can try to fix that though, but it's hard to test.

The windows version seems to be functioning. You have to run it once, then it registers itself for the sc4pac protocol, and subsequently you can open links of the type [sc4pac://memo.github.io/sc4pac/channel/?pkg=memo:submenus-dll](sc4pac://memo.github.io/sc4pac/channel/?pkg=memo:submenus-dll), both when the app is running or not.

**Comment by sebamarynissen** [comment]
New build with a version that might be working on Linux too: https://github.com/sebamarynissen/sc4pac-protocol/actions/runs/12611183032

**Comment by memo33** [comment]
Thanks for the PoC. For me, the main question isn't even whether it can be done – it's actually doing it. Flutter does provide ways to use custom protocols, too ([app_links](https://pub.dev/packages/app_links)).

The intention behind the web-app is to offer it for people who cannot run the native apps. However, that's where dealing with custom protocols becomes difficult. Wrapping the web-app in Electron doesn't really make sense, as it defeats the point of choosing the web-app over the native Flutter apps.

The reason I currently don't build a native macOS app is that Apple made it a hard requirement to notarize your apps (which requires an Apple developer license), at least according to several online sources on the topic. I'm not 100% sure though what happens if you distribute an unnotarized application. There might still be ways for users to run it if they disable some security checks, but I vaguely remember reading that it would be limited to 100 users or so, being intended for development purposes. In any case, the same limitation would apply to Electron apps, so I'd prefer to use the Flutter approach rather than Electron.

**Comment by sebamarynissen** [comment]
Turns out it wasn't working after all. I was accidentally relying on an instance of the sc4pac server that I left running, but it wasn't correctly started up in the application. I fixed it and polished the startup procedure a bit, so it should work now: https://github.com/sebamarynissen/sc4pac-protocol/actions/runs/12611990868

I understand not wanting to go with the Electron approach. However, I have created Electron apps for Mac in the past, and I never needed an Apple developer license. Not sure if this holds for the way Flutter builds a native app though. Do you have a Mac at your disposal? I could build a Mac version of the protocol PoC as well and then we can see what happens.

**Comment by memo33** [comment]
I don't have a Mac with a modern version of macOS, I'm afraid. It's only recent versions of the OS in which Apple has hardened their requirements more and more.

Though, I've found a [recent source](https://www.iclarified.com/94521/apple-will-make-it-harder-to-run-unnotarized-apps-in-macos-sequoia) which sounds like it would still be possible to run unnotarized applications, as long as users explicitly bypass Gatekeeper.

**Comment by sebamarynissen** [comment]
My wife has a 2024 Mac. I'll ask her politely and see if I can test an Electron-wrapper sc4pac-gui on it. If you have the ability to build a Mac version of the Flutter version as well, that'd be great, then I can test it too.

**Comment by memo33** [comment]
That would be awesome. I've run a test build for macOS here: https://github.com/memo33/sc4pac-gui/actions/runs/12612840607. The attached [artifact file](https://github.com/memo33/sc4pac-gui/actions/runs/12612840607/artifacts/2386157996) is available for 7 days. It includes a .tar file from which you only need the `sc4pac-gui.app` subfolder I think.

The main question is whether the application is runnable. It doesn't include an implementation for a custom protocol yet. The .app folder also doesn't yet include the CLI, as I need to do some research first on where to put the `cli` folder inside the .app folder. This means you'd either need to run the `sc4pac server` separately or use the `--sc4pac-cli-dir <path>` launch parameter for the GUI. In the latter case, you might also need to pass `--profiles-dir <path>`, as the default location would be inside the .app folder and I'm not sure if the application has write permissions there. (I'm considering to change the default `profiles` location as part of #16.)

**Comment by sebamarynissen** [comment]
Looks like it's not working. When I try to open it, MacOS tells me it can't very that it isn't malware and only offers me to move it to the recycle bin. I'm going to try with a basic Electron app and see if the same happens in that case. If that works, the Electron approach could at least be a solution for Mac.

**Comment by memo33** [comment]
Have you tried this tip from the article I referenced above?

> Users will now need to visit System Settings > Privacy & Security to review security information for software before allowing it to run.

**Comment by sebamarynissen** [comment]
Then it allows me to run it, but it tries to super scary. It mentions 3 times that it can't verify it's not malware and then asks for the password. It crashes immediately on startup though.

**Comment by memo33** [comment]
Yeah, that's the dilemma with Apple nowadays. I'm not sure if the average user will be trained to bypass Gatekeeper or to use no third-party software. According to [this source](https://bigbinary.com/blog/code-sign-notorize-mac-desktop-app), it sounds like an Electron app would face the same problem.

If there are crashes, could you post the trace please?

**Comment by sebamarynissen** [comment]
Jfc, this new policy is annoying. You can't do anything anymore. Trying to get the cli to run requires me to perform the same steps: manually verifying it in the System Security Settings. What an absolute disgrace. Another blatant effort by Darth Cook to try and take over control "in the interest of our users", just like they don't allow you to do anything with your iPhone that you own ffs.

So yeah, the Electron application is going to face the same problems indeed. I didn't know it was such a recent change.

**Comment by memo33** [comment]
Dang it, that's too bad. Somehow I was hoping that the CLI wasn't affected by this, as the actual application behind it would just be "Java" which would be installable from official sources. Thank you for testing this.

**Comment by sebamarynissen** [comment]
I got the cli running now, but the crashes remain. I don't really see any logs or traces, and I'm too inexperienced with Mac to hunt this down. For me, the fact that we will have to pay $99 a year for a game that doesn't even run on a Mac (I know about the Aspyr port, but that's no longer supported now by ST anyway) ends it for me. I don't really see the point of putting a lot of efforts in getting to run this on Mac.

Anyway, is there anything else holding you back from trying to register the gui as a protocol handler for sc4pac://? The experience I had when testing it out with the Electron app is really, really nice, as sc4pac opens up when clicking the links, even when it's not running yet.

**Comment by memo33** [comment]
Fair point. You're probably right we shouldn't be too concerned about Mac support. I do know of one user who's successfully used the CLI on macOS before. Even when you run the Windows version of the game using Wine or in a VM, it makes sense to be able to use sc4pac natively. If the custom protocol is the only thing that doesn't work on Mac, so be it, I guess.

> Anyway, is there anything else holding you back from trying to register the gui as a protocol handler for sc4pac://? The experience I had when testing it out with the Electron app is really, really nice, as sc4pac opens up when clicking the links, even when it's not running yet.

No. I'll try to look into it more in the coming days.

**Comment by sebamarynissen** [comment]
If you need help with testing the registry editing script on Windows, let me know.

One last thing that I still wanted to say about the Electron wrapper approach, and then I'll shut up about it. Using it has a few other advantages on top of the automatic protocol registration. It would also allow us to make use of the [builtin auto-updating features](https://www.electronjs.org/docs/latest/tutorial/updates).

It also opens up the possibility to automatically handle the authentication, as I've already tried out in https://github.com/memo33/sc4pac-tools/issues/9#issuecomment-2559632917. I know you think it's not the way to go, but this way it could be "integrated" in the actual sc4pac application, so maybe that makes it more acceptable. I have several applications where I have to login with the same account credentials as on the website (Steam, Gog, ...), although the difference there is obviously that the application is closely tied to the website itself, which is not really the case with sc4pac of course.

**Comment by memo33** [comment]
Closing this as "not planned", as the blocking of localhost requests is ultimately a client issue, which cannot be fixed from the sc4pac side.

Workarounds:
- using the `sc4pac://` protocol buttons with the sc4pac GUI desktop app (unsupported by the web-app)
- using a different browser
- or configuring the browser as described in https://github.com/memo33/sc4pac-gui/issues/15#issuecomment-2569697672 above

> It would also allow us to make use of the [builtin auto-updating features](https://www.electronjs.org/docs/latest/tutorial/updates).

The web-app has some drawbacks compared to the desktop app, especially performance and it lacks a few features that aren't available inside the web browser. So the web-app is merely a fallback solution. Being a package manager, an auto-update feature of the application itself could probably be implemented, but it's not too important at this point.
